File: Homework-by-C++/Homework-Week-4/Hw 4_3 大師と仙人との奇遇.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

long long calculate_profit(int N, vector<int>& prices) {
    priority_queue<int, vector<int>, greater<int>> min_heap;  // Min-heap to track the prices of stocks bought
    long long total_profit = 0;

    for (int i = 0; i < N; ++i) {
        // Buy a stock every day
        min_heap.push(prices[i]);

        // If it's not the last day, check if there are any stocks to sell
        if (i != N - 1) {
            while (!min_heap.empty() && min_heap.top() < prices[i + 1]) {
                total_profit += prices[i + 1] - min_heap.top();
                min_heap.pop();
            }
        } 
        // On the last day, sell all remaining stocks
        else {
            while (!min_heap.empty()) {
                total_profit += prices[i] - min_heap.top();
                min_heap.pop();
            }
        }
    }
    return total_profit;
}

int main() {
    int N;
    cin >> N;
    vector<int> prices(N);
    for (int i = 0; i < N; ++i) {
        cin >> prices[i];
    }

    cout << calculate_profit(N, prices) << endl;
    return 0;
}

File: Homework-by-python/Homework-Week-4/Hw 4_3 大師と仙人との奇遇.py
import heapq

def calculate_profit(N, prices):
    # Using a list to represent the min-heap for bought stock prices
    bought_prices = []
    # Total profit made
    total_profit = 0

    for i in range(N):
        # Buy a stock every day
        heapq.heappush(bought_prices, prices[i])

        # If it's not the last day, check if there are any stocks to sell
        if i != N - 1:
            # Check the stock with the lowest purchase price
            while bought_prices and bought_prices[0] < prices[i + 1]:
                # Update the total profit by selling the stock
                total_profit += prices[i + 1] - heapq.heappop(bought_prices)
        
        # On the last day, sell all remaining stocks in hand
        else:
            while bought_prices:
                total_profit += prices[i] - heapq.heappop(bought_prices)

    return total_profit

# Input
N = int(input())
prices = list(map(int, input().split()))

# Output
print(calculate_profit(N, prices))
